extra/onnx_ops.py

```diff
@@ -601,15 +609,18 @@ def AffineGrid(theta: Tensor, size: Tensor, align_corners=0):
 
 # **************** com.microsoft Ops ****************
 
+@half_to_float_to_half
 def SkipLayerNormalization(x:Tensor, skip:Tensor, gamma, beta:Optional[Tensor]=None, bias:Optional[Tensor]=None, epsilon=None):
-  if epsilon is None: epsilon=1e-12
+  if epsilon is None: epsilon = 1e-12 if x.dtype == dtypes.float else 1e-5 # Tensor can only be float or half
   x = x + skip + bias
   return x.layernorm(eps=epsilon) * gamma + beta, None, None, x
 
+@half_to_float_to_half
 def FastGelu(x:Tensor, bias:Optional[Tensor]=None):
   x = x + bias
   return 0.5 * x * (1 + (x * 0.797885 + 0.035677 * x ** 3).tanh())
 
+@half_to_float_to_half # l2 norm for commavq fails without this
```

Comment:
it turns out `commavq` also has some shocking float16 weirdness
the inaccuracy here is large.. (like 20.0 to 0.15 kind of large)
My guess is some op is making these float16 numbers go under the threshold, and it ends up as 0.0

need to investigate.

Reply:
i think that's where we need to update onnx implementation. Normalization in half in brittle so we usually cast training pipeline to float before batchnorm for example